app/javascript/extensible-components/index.ts

```diff
@@ -0,0 +1,67 @@
+import { IExtensionComponent, IMiQApiCallback } from './lib';
+
+/**
+ * Class for easy creation of extensible component.
+ */
+export class ExtensibleComponent {
+  public unsubscribe: Function;
+  constructor(public name: string, public api: IMiQApiCallback, public render: IMiQApiCallback){}
+}
+
+/**
+ * Create new object which will hold extension components on MiQ main object.
+ */
+ManageIQ.extensionComponents = ManageIQ.extensionComponents || {};
+
+/**
+ * Subject from from Rxjs to send message that we want to register new component.
+ */
+ManageIQ.extensionComponents.source = new Rx.Subject();
+
+/**
+ * Components will be saved in items which will be set. To easy remove of components which no longer exists.
+ */
+ManageIQ.extensionComponents.items = new Set();
+
+/**
+ * Helper function to create new component.
+ * @param name string name of new component.
+ * @param api callback functions to change inner logic of component.
+ * @param render callback function to apply render functions.
+ */
+ManageIQ.extensionComponents.newComponent = function(name: string, api?: IMiQApiCallback, render?: IMiQApiCallback) {
+  let newCmp = new ExtensibleComponent(name, api, render);
+  ManageIQ.extensionComponents.source.onNext({action: 'add', payload: newCmp});
+  return newCmp;
```

Comment:
Do we really need the return value, and if so, why?

Reply:
When creating and adding new component new component will also has `unbind` function to deregister itself.